### PR TITLE
Clean up delphi_utils (take 2)

### PR DESCRIPTION
**Pull request by melange396**
### Description
migrating setup.py to pyproject.toml and cleaning up dependencies

### Changelog
Itemize code/test/documentation changes and files added/removed.
- added pyproject.toml

### Associated Issue(s) 
- Addresses #1973 

(this is a redo of https://github.com/cmu-delphi/covidcast-indicators/pull/1984, which was reverted in #2001 due to problems with Jenkins builds)

### CODE REVIEW

File: changehc/Makefile [hunk 1]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline changehc/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: claims_hosp/Makefile [hunk 2]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline claims_hosp/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: doctor_visits/Makefile [hunk 3]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline doctor_visits/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: google_symptoms/Makefile [hunk 4]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline google_symptoms/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: hhs_hosp/Makefile [hunk 5]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline hhs_hosp/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: nchs_mortality/Makefile [hunk 6]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline nchs_mortality/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: nssp/Makefile [hunk 7]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline nssp/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: nwss_wastewater/Makefile [hunk 8]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline nwss_wastewater/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: quidel_covidtest/Makefile [hunk 9]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline quidel_covidtest/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: quidel_covidtest/setup.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 from setuptools import find_packages
 
 required = [
+    "boto3",
```
Comment (dshemetov) [inline quidel_covidtest/setup.py:5]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

File: sir_complainsalot/Makefile [hunk 11]
```diff
@@ -13,7 +13,7 @@ install: venv
 install-ci: venv
 	. env/bin/activate; \
 	pip install wheel ; \
-	pip install ../_delphi_utils_python ;\
+	pip install '../_delphi_utils_python[dev]' ;\
```
Comment (dshemetov) [inline sir_complainsalot/Makefile:16]:
```suggestion
	pip install ../_delphi_utils_python ;\
```

**Review by dshemetov — CHANGES_REQUESTED** [review]
I didn't catch in the previous PR that the individual indicators don't actually need the dev version of the delphi utils, so made those suggestions. If the indicators have dependency issues because of these changes, we should add the dependencies to the indicator's dependency list (seems like good form not to have packages depend on other packages dev dependencies).

File: _delphi_utils_python/pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    "cvxpy",
+    "epiweeks",
+    "gitpython",
+    "importlib_resources>=1.3",
+    "numpy",
+    "pandas>=1.1.0",
+    "slackclient",
+    "scs<3.2.6", # TODO: remove this ; it is a cvxpy dependency, and the excluded version appears to break our jenkins build. see: https://github.com/cvxgrp/scs/issues/283
+    "structlog",
+    "xlrd",
```
Comment (dshemetov) [inline _delphi_utils_python/pyproject.toml:30]:
```suggestion
    "pandas[excel]>=1.1.0",
    "requests",
    "slackclient",
    "scs<3.2.6", # TODO: remove this ; it is a cvxpy dependency, and the excluded version appears to break our jenkins build. see: https://github.com/cvxgrp/scs/issues/283
    "structlog",
```
Suggestion from [comment here](https://github.com/cmu-delphi/covidcast-indicators/issues/1973#issuecomment-2272154545).

File: google_symptoms/Makefile (hunk 4 above)
Comment (aysim319) [reply google_symptoms/Makefile:16]: Has an interesting error; when I don't do the [dev] it's saying that covidcast doesn't have usi_api_key as a module???

`delphi_google_symptoms/date_utils.py:77:4: E1101: Module 'covidcast' has no 'use_api_key' member (no-member)`

https://github.com/cmu-delphi/covidcast-indicators/actions/runs/10285726380/job/28464696528

File: sir_complainsalot/Makefile (hunk 11 above)
Comment (aysim319) [reply sir_complainsalot/Makefile:16]: Here too: the linter fails when I took the 'dev' off

https://github.com/cmu-delphi/covidcast-indicators/actions/runs/10285993553/job/28465604873?pr=2014

**Comment by aysim319** [comment]
> I didn't catch in the previous PR that the individual indicators don't actually need the dev version of the delphi utils, so made those suggestions. If the indicators have dependency issues because of these changes, we should add the dependencies to the indicator's dependency list (seems like good form to isolate dev dependencies).

most of these need for mock, the only other package that has mock as a requirement in setup.py is google symptoms.

File: quidel_covidtest/setup.py (hunk 10 above)
Comment (aysim319) [reply quidel_covidtest/setup.py:5]: ? is this a typo?

**Comment by aysim319** [comment]
Do we also want to convert the rest of the indicator to pyproject.toml? maybe putting in the template_python to starting thing off?

**Comment by dshemetov** [comment]
> most of these need for mock

Right, I'm saying we should add mock to each indicator's individual setup.py instead of grabbing it from _delphi_utils' dev dependencies. If you feel up for migrating the indicators to pyproject.toml, then each indicator can have its own "dev" set of packages.

Also, please add `requests` to the _delphi_utils `pyproject.toml`, it was in my suggestion above.

File: quidel_covidtest/setup.py (hunk 10 above)
Comment (dshemetov) [reply quidel_covidtest/setup.py:5]: ah yea, my bad

**Comment by aysim319** [comment]
Oh I didn't see it before 🤦‍♀️Thanks for pointing it out

**Comment by dshemetov** [comment]
FYI, [this Jenkins script](https://github.com/cmu-delphi/covidcast-indicators/blob/main/jenkins/build-indicator.sh) will need to be adjusted to check for the presence of `pyproject.toml` instead.

**Comment by aysim319** [comment]
> > most of these need for mock
> 
> Right, I'm saying we should add mock to each indicator's individual setup.py instead of grabbing it from _delphi_utils' dev dependencies. Sounds good on migrating the indicators to pyproject.toml, then each indicator can have its own "dev" set of packages and we can put `mock` in there.

Okay just making sure I'm understanding this properly; do we want to convert the existing indicators also to pyproject.toml with the same setup as delphi_utils that has an optional requirement in this PR?

**Comment by dshemetov** [comment]
I don't have a strong preference. Getting delphi_utils cleaned is the key task here and this PR is close to taking care of that (I still think we should remove the ".[dev]" from the individual indicators and debug this CI). Porting the indicators to `pyproject.toml` is nice to have for consistency, but not critical.

**Review by dshemetov — APPROVED** [review]
This is good to go. Let's make an issue to migrate the indicators to pyproject.toml and link to this thread. Let's remember to squash merge, so the commit log gets cleaned.

**Pull request merged**